### ISSUE TITLE
not working with ol 4.5 in webpack

### ISSUE THREAD

**Issue by niteshgoyal19** [jonataswalker/ol-geocoder#152]
gives an error of append child of undefined while adding Control to openlayer

**Comment by jonataswalker** [comment]
@niteshgoyal19 Are you able to provide a repro? Like a fiddle.

**Comment by niteshgoyal19** [comment]
Not now but u can try if u use angular 5 ol is coming in different module u
have to import those module so in that case geocoder object not added as
control in ol

On 30 Nov 2017 3:04 pm, "Jonatas Walker" <notifications@github.com> wrote:

@niteshgoyal19 <https://github.com/niteshgoyal19> Are you able to provide a
repro? Like a fiddle.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/jonataswalker/ol-geocoder/issues/152#issuecomment-348132866>,
or mute the thread

**Comment by jonataswalker** [comment]
Sorry, don't have time to build this environment and discover what's happening.

**Comment by aguynamedben** [comment]
Thanks for the time you spend working on ol-geocoder. Like @jonataswalker and @caseycesari I'm also running into this issue.

![image](https://user-images.githubusercontent.com/107841/34633751-1a2c4704-f233-11e7-86da-c682ecd42192.png)

The stacktrace points to ./node_modules/openlayers/dist/ol.js:142. My project is attempting to use ol 4.6.4 via the modern ES2015+Webpack method linked to in the docs [here](https://gist.github.com/tschaub/79025aef325cd2837364400a105405b8). This is the only part of my project that is failing with Webpack.

I'll try to reproduce in a test repo so you can see what's going on. It seems that the ol/openlayers npm packages recommend using the `ol` npm package for Webpack/Bowerify/ES6 usage, but using `openlayers` for pre-built usage. My project is trying to use `ol`, and it looks like ol-geocoder depends on `openlayers`. Maybe there is some discrepancy with how the imports are happenings vs. the dist/ version. Ideally, I'd like none of my dependencies to use a dist/ version of OpenLayers. I want Webpack to package up the ES6 modules as needed.

@jonataswalker @caseycesari were y'all able to find any temporary workarounds?

I'll post here when I have a repo with a reproduction.

**Comment by aguynamedben** [comment]
I'm curious if this issue would be fixed by depending on `ol` instead of `openlayers`, then changing the imports to the recommended ES6 way.

I.e. go from this style:
```js
import ol from 'openlayers';

const layer = new ol.layer.Vector({
  name: layerName,
  source: new ol.source.Vector()
});
```

To this:
```js
import SourceVector from 'ol/source/vector';
import LayerVector from 'ol/layer/vector';

const layer = new LayerVector({
  name: layerName,
  source: new SourceVector()
});
```

I'm an OpenLayers n00b, but from my read of the docs, that's how OpenLayers should be used in an ES6 environment.

Will reproduce, I'm curious what the maintainers think of that, though!

**Comment by aguynamedben** [comment]
@jonataswalker I was able to reproduce in this small, simple repo: https://github.com/aguynamedben/ol-geocoder-152 details in the README. I'm curious what you think. Thanks for your time.

**Comment by walkermatt** [comment]
I've been doing some work moving the ol3-layerswitcher to use `ol` package, see https://github.com/walkermatt/ol3-layerswitcher/issues/69. I think all plugins are going to need to move over to the `ol` package shortly.

For ol-geocoder I think the change should be fairly simple; in the source import the individual classes from `ol` instead of `openlayers` then update rollup build to mark each ol class external and define globals for each similar to https://github.com/walkermatt/ol3-layerswitcher/blob/ol/rollup.config.js#L20-L28.

**Comment by jonataswalker** [comment]
Yup, working on it.

**Comment by aguynamedben** [comment]
🙏 Thank you for looking at it. I'm not well-versed in OpenLayers or ol-geocoder, but I'm happy to help. I think I understand the process of moving to the `ol` package because I had to do it in my project.

**Comment by jonataswalker** [comment]
I'm stuck and I'll need some help, @walkermatt?

There are three scenarios:

1. `ol` package dependency
   * works on browser
   * works on project that depends on `ol`
   * doesn't work on unit tests (`test:unit`)
      * unit tests are done in a Node environment therefore `SyntaxError: Unexpected token import`
2. no dependencies at all
   * works on browser
   * doesn't work on project that depends on `ol`
      * `ol` (used by `ol-geocoder`) is `undefined`
   * works on tests
      * unit tests are done with `openlayers` package
3. `openlayers` package dependency
   * the reason this issue was created

**Comment by walkermatt** [comment]
I think dropping support for the `openlayers` package in a major version is fine which leaves support for: 
1. `ol` package
2. hosted build (`ol.Map` etc.)

In ol3-layerswitcher I'm building a [UMD bundle which supports the `ol` package](https://github.com/walkermatt/ol3-layerswitcher/blob/ol/dist/ol-layerswitcher.js) and [the hosted build](https://github.com/walkermatt/ol3-layerswitcher/blob/ol/src/ol3-layerswitcher.js#L294-L298).

Currently the tests use the hosted build.

**Comment by aguynamedben** [comment]
@jonataswalker Any update on migrating to the `ol` package? I noticed some other plugins have done this now (https://github.com/walkermatt/ol-popup/commit/494a42c04d27319ef95e9c5abdd34f1289b43a72).

Is it a matter of simply switching the imports in src/nominatim.js and src/base.js and changing the dependencies? I'm happy to take a stab at it if you think it's pretty straightforward.

**Comment by jonataswalker** [comment]
Fixed in #162. Thanks @aguynamedben @walkermatt
